File: lib/conway_deathmatch.ex
defmodule ConwayDeathmatch do
  @module_doc """
  Provides Conway's Game of Life

  Significant features:
  * Deathmatch - multiple competing populations
  * Several rulesets:
    - Traditional - Single population
    - Aggressive - Alive cells can switch sides
    - Defensive - Alive cells never switch sides
    - Friendly - Only friendly cells matter
  * Toroidal grid - the grid wraps left/right and up/down
  * Configurable render representation - uses @dead, and any non-@dead values
                                         identify a specific (alive) population
"""
  @dead "."
  defstruct grid: [[]], width: 0, height: 0,
  dead: @dead, deathmatch: nil, ticks: 0

  def new(width, height) when is_integer(width) and width > 0 and is_integer(height) and height > 0 do
    grid = for _y <- 0..(height - 1) do
      for _x <- 0..(width - 1) do
        @dead
      end
    end
    %ConwayDeathmatch{grid: grid, width: width, height: height}
  end
  def new(width, height, points, val \\ 0) do
    new(width, height) |> add_points(points, val)
  end

  def add_points(conway, points, val) do
    grid = for y <- 0..(conway.height - 1) do
      for x <- 0..(conway.width - 1) do
        if HashSet.member?(Enum.into(points, HashSet.new), [x, y]) do
          val
        else
          conway |> cell(x, y)
        end
      end
    end
    %ConwayDeathmatch{conway | grid: grid}
  end

  def cell(conway, x, y) do
    conway.grid |> Enum.at(y) |> Enum.at(x)
  end

  def tick(conway) do
    grid = for y <- 0..(conway.height - 1) do
      for x <- 0..(conway.width - 1) do
        conway |> next_cell(x, y)
      end
    end
    %ConwayDeathmatch{conway | grid: grid, ticks: conway.ticks + 1}
  end

  def render(conway) do
    conway.grid |> Enum.map(&Enum.join/1) |> Enum.join("\n")
  end
  def print(conway, %{renderfinal: _rf}), do: conway
  def print(conway, _options) do
    IO.puts ""
    IO.puts conway.ticks
    IO.puts render(conway)
    conway
  end

  # TODO: switch on deathmatch
  def next_cell(conway, x, y) do
    cur = conway |> cell(x, y)
    case conway.deathmatch do
      _ -> cur |> next_cell_sp(neighbors(conway, x, y))
    end
  end

  # traditional, single population
  def next_cell_sp(cell_val, neighbors) do
    living = neighbors |> alive
    case {cell_val, length(living)} do
      {@dead, 3} -> hd(living)
      {v, 2} -> v
      {v, 3} -> v
      {_, _} -> @dead
    end
  end

  def neighbors(conway, x, y) do
    for xn <- (x-1)..(x+1), yn <- (y-1)..(y+1), {xn,yn} != {x,y} do
      cell conway, tor(xn, conway.width), tor(yn, conway.height)
    end
  end

  def alive(neighbors) do
    neighbors |> Enum.reject(fn n -> n == @dead end)
  end

  def group(neighbors) do
    Enum.group_by(neighbors, &(&1))
  end

  def majority([head | tail]) do
    Enum.reduce(tail, {head, Dict.put(%{}, head, 1)},
      fn i, {lead, group} ->
        i_count = Dict.get(group, i, 0) + 1
        if i_count > Dict.get(group, lead) do
          {i, Dict.put(group, i, i_count)}
        else
          {lead, Dict.put(group, i, i_count)}
        end
      end)
    |> elem(0)
  end

  # toroidal   tor(-2, 5) == 3
  def tor(num, mod) when is_integer(num) and is_integer(mod) and mod > 0 do
    case rem(num, mod) do
      r when r >= 0 -> r
      r             -> r + mod
    end
  end

  def default_options do
    %{width: 40,
      height: 20,
      shapes: "p 2 0 p 3 0 p 4 0 p 5 0 p 6 0 p 7 0 p 8 0 p 9 0 p 10 0",
      sleep: 0}
  end

  def loop(conway, options \\ default_options()) do
    conway |> tick
    |> print(options) |> stop(options) |> sleep(options)
    |> loop(options)
  end

  def stop(conway, %{ticks: ticks}) do
    if conway.ticks >= ticks, do: System.halt(0), else: conway
  end
  def stop(conway, _options), do: conway

  def sleep(conway, %{sleep: s}) when s == 0, do: conway
  def sleep(conway, %{sleep: s}), do: :timer.sleep(s) && conway
end

File: lib/conway_deathmatch.rb
# Provides a 2d array for the grid
# Implements standard and deathmatch evaluation rules
# Boundaries are toroidal: they wrap in each direction
#
class ConwayDeathmatch
  DEAD = '.'
  ALIVE = '0'

  def self.new_grid(width, height)
    grid = []
    width.times { grid << Array.new(height, DEAD) }
    grid
  end

  # nil for traditional, otherwise :aggressive, :defensive, or :friendly
  attr_accessor :deathmatch

  def initialize(width, height, deathmatch = nil)
    @width = width
    @height = height
    @grid = self.class.new_grid(width, height)
    @deathmatch = deathmatch
  end

  # Conway's Game of Life transition rules
  def next_value(x, y)
    # don't bother toroidaling, only called by #tick
    n, birthright = neighbor_stats(x, y)
    if @grid[x][y] != DEAD
      (n == 2 or n == 3) ? birthright : DEAD
    else
      (n == 3) ? birthright : DEAD
    end
  end

  def value(x, y)
    @grid[x % @width][y % @height]
  end

  # total (alive) neighbor count and birthright
  def neighbor_stats(x, y)
    x = x % @width
    y = y % @height
    npop = neighbor_population(x, y).tap { |h| h.delete(DEAD) }

    case @deathmatch
    when nil
      [npop.values.reduce(0, :+), ALIVE]

    when :aggressive, :defensive
      # dead: determine majority (always 3, no need to sample for tie)
      # alive: agg: determine majority (may tie at 2); def: cell_val
      determine_majority = (@grid[x][y] == DEAD or @deathmatch == :aggressive)
      total = 0
      largest = 0
      birthrights = []
      npop.each { |sym, cnt|
        total += cnt
        return [0, DEAD] if total > 3   # [optimization]
        if determine_majority
          if cnt > largest
            largest = cnt
            birthrights = [sym]
          elsif cnt == largest
            birthrights << sym
          end
        end
      }
      [total, determine_majority ? (birthrights.sample || DEAD) : @grid[x][y]]

    when :friendly
      # [optimization] with knowledge of conway rules
      # if DEAD, need 3 friendlies to qualify for birth sampling
      # if ALIVE, npop simply has the friendly count
      cell_val = if @grid[x][y] == DEAD
                   npop.reduce([]) { |memo, (sym,cnt)|
                     cnt == 3 ? memo + [sym] : memo
                   }.sample || DEAD
                 else
                   @grid[x][y]
                 end
      # return [0, DEAD] if no one qualifies
      [npop[cell_val] || 0, cell_val]
    else
      raise "unknown: #{@deathmatch.inspect}"
    end
  end

  # population of every neighboring entity, including DEAD
  def neighbor_population(x, y)
    x = x % @width
    y = y % @height
    neighbors = Hash.new(0)
    (x-1).upto(x+1) { |xn|
      (y-1).upto(y+1) { |yn|
        xn = xn % @width
        yn = yn % @height
        neighbors[@grid[xn][yn]] += 1 unless (xn == x and yn == y)
      }
    }
    neighbors
  end

  # generate the next grid table
  def tick
    new_grid = self.class.new_grid(@width, @height)
    @width.times { |x|
      @height.times { |y| new_grid[x][y] = next_value(x, y) }
    }
    @grid = new_grid
    self
  end

  # set a single point
  def populate(x, y, val = ALIVE)
    @grid[x % @width][y % @height] = val
  end

  # set several points (2d array)
  def add_points(points, x_off = 0, y_off = 0, val = ALIVE)
    points.each { |point| populate(point[0] + x_off, point[1] + y_off, val) }
    self
  end

  # for line-based text output, iterate over y-values first (i.e. per row)
  def render_text
    @grid.transpose.map { |row| row.join }.join("\n")
  end
  alias_method :render, :render_text

  # full grid scan
  def population
    population = Hash.new(0)
    @grid.each { |col| col.each { |val|  population[val] += 1  } }
    population
  end
end
